anndata/compat/__init__.py

```diff
@@ -62,12 +62,37 @@ class Literal(metaclass=LiteralMeta):
             pass
 
 
+def _read_hdf5_attribute(attrs: h5py.AttributeManager, name: str):
+    """
+    Read an HDF5 attribute and perform all necessary conversions.
+
+    At the moment, this only implements conversions for string attributes, other types
+    are passed through. String conversion is needed compatibility with other languages.
+    For example Julia's HDF5.jl writes string attributes as fixed-size strings, which
+    are read as bytes by h5py.
+    """
+    attr = attrs[name]
+    attr_id = attrs.get_id(name)
+    dtype = h5py.check_string_dtype(attr_id.dtype)
+    if dtype is None:
+        return attr
+    else:
+        if dtype.length is None:  # variable-length string, no problem
+            return attr
+        elif len(attr_id.shape) == 0:  # Python bytestring
+            return attr.decode("utf-8")
+        else:  # NumPy array
+            return attr.astype("U")
```

Comment:
We generally don't use the numpy unicode type, instead opting for just arrays of python strings. The numpy unicode type is really weird (utf-32, not supported by h5py), and we haven't really had a use for fixed length text data. Also numpy just doesn't have good tools for working with text data, generally we hand this to pandas which uses object arrays.

Reply:
The problem is that you can get fixed-length strings from HDF5 files:
```
f.attrs.create("test", np.asarray(["test1", "test2"]), dtype=h5py.h5t.string_dtype(length=15))
f.attrs["test"]
```
gives 
```
array([b'test1', b'test2'], dtype='|S15')
```
and now it needs to be converted to something that we can work with. `np.astype("O")` doesn't work here, and you do use `np.astype("U")` in a couple of places, e.g. [here](https://github.com/theislab/anndata/blob/master/anndata/utils.py#L151) [here](https://github.com/theislab/anndata/blob/master/anndata/compat/__init__.py#L82) [here](https://github.com/theislab/anndata/blob/master/anndata/_core/merge.py#L783) (the last of which, `astype(str)` also results in a `U` dtype)

Reply:
> The problem is that you can get fixed-length strings from HDF5 files

We can get fixed length strings from hdf5 files, but we won't keep them as fixed length strings in memory. I'd say our goal is not to have a one-to-one mapping with hdf5, but to be able to represent useful data types. There isn't really a useful fixed length text type in our dependencies or a use case for it, so we don't have a representation for it.

> and now it needs to be converted to something that we can work with. np.astype("O") doesn't work here

This should work in this case:

```python
import codecs, h5py

string_type = h5py.check_string_type(attrs.get_id(key).dtype)
[codecs.decode(x, string_type.encoding) for x in attrs[key]]
```

Passing the encoding isn't even necessary, since hdf5 only uses ascii and utf-8 so utf-8 should always work. So just

```python
[codecs.decode(x) for x in attrs[key]]
```

> in a couple of places, e.g. here here here

As for those examples:

* `convert_dictionary_to_structured_array` is never called as far as I can tell. We've also moved significantly away from structured arrays since then. I also think this code predates me working on AnnData.
* `_from_fixed_length_strings` is specifically for backwards compatibility. The main goal of that function is to convert fixed length strings to python `str`s anyways. We definitely would not recommend putting anything in structured arrays these days, as they are such a massive pain to work with.
* `np.arange(len(adatas)).astype(str)` I hadn't realized this was fixed length, but I'm not sure it ultimately matters since the only way it's getting into the `AnnData` object is through

```python
    label_col = pd.Categorical.from_codes(
        np.repeat(np.arange(len(adatas)), [a.shape[axis] for a in adatas]),
        categories=keys,
    )
```

Where:

```python
label_col.categories.dtype
# dtype('O')
```

Could totally be replaced with `keys = [str(k) for k in range(len(adatas))]`.